### PR TITLE
TURN_ASSIST pitch gain for coordinated turns

### PR DESCRIPTION
**Pull request by avsaase**
This PR implements `fw_turn_assist_pitch_gain` for the amount of pitch to maintain a constant pitch angle during coordinated turns in `TURN ASSIST`. This is similar to the already existing `fw_turn_assist_yaw_gain`. I have tested this in flight and it works as expected.

### CODE REVIEW

File: src/main/flight/pid.h [hunk 1]
```diff
@@ -126,6 +126,7 @@ typedef struct pidProfile_s {
     uint16_t    fixedWingItermThrowLimit;
     float       fixedWingReferenceAirspeed;     // Reference tuning airspeed for the airplane - the speed for which PID gains are tuned
     float       fixedWingCoordinatedYawGain;    // This is the gain of the yaw rate required to keep the yaw rate consistent with the turn rate for a coordinated turn.
+	float       fixedWingCoordinatedPitchGain;    // This is the gain of the pitch rate to keep the pitch angle constant during coordinated turns.
```
Comment (digitalentity) [inline src/main/flight/pid.h]: Alignment problem here

File: src/main/flight/pid.c [hunk 2]
```diff
@@ -256,6 +256,7 @@ PG_RESET_TEMPLATE(pidProfile_t, pidProfile,
         .fixedWingItermThrowLimit = FW_ITERM_THROW_LIMIT_DEFAULT,
         .fixedWingReferenceAirspeed = 1000,
         .fixedWingCoordinatedYawGain = 1.0f,
+		.fixedWingCoordinatedPitchGain = 1.0f,
```
Comment (digitalentity) [inline src/main/flight/pid.c]: Alignment problem here (probably \t used instead of 4-spaces)

**Review by digitalentity — COMMENTED** [review]
I'm not sure if this is needed at all. Required amount of pitch/yaw is calculated based on airspeed and bank angle. If the plain needs more pitch, then the airspeed is likely not correct. This parameter gives a false perception than it's safe to leave airspeed sensor/parameter incorrect and tweak only this one.

**Comment by avs2** [comment]
I made this PR because I had a problem with my Dart 250 pitching up and gaining altitude when making turns in the navigation modes. See my post on the RCGroups thread [here](https://www.rcgroups.com/forums/showpost.php?p=45357477&postcount=26435). Naturally I tried to reduce pitch FF, but that cost me so much elevator authority that the plane became barely flyable. I tried a bunch of different settings for rates, piffs and mixer weights, none of which completely solved the problem, and all made the plane fly worse in acro and angle. I agree that in theory this PR should be unnecessary, but in practice for me it is not. 

As an aside, is the normal piff controller not already perfectly capable to keeping the pitch angle at zero?

**Comment by digitalentity** [comment]
> Also, most users will probably never hear of this setting, just like they've never heard of the reference airspeed and the yaw gain setting. Then there is no harm done by including it.

Makes sense, default of `1` changes nothing for users.

> As an aside, is the normal piff controller not already perfectly capable to keeping the pitch angle at zero?

In theory - yes, however, ANGLE controller is **reactive** in nature, meaning it will only act when there is an angle error. Turn-assist acts **proactive** preventing the error from occurring, rather than correcting the error which has already developed.

**Comment by avs2** [comment]
Yesterday I experimented again with this PR and `fw_turn_assist_pitch_gain = 0` gave me nice zero pitch, constant altitude turns in the navigation modes with. So for me at least the airplane flies much better with this code, and I hope this makes it into 2.6.

I also accidentally experimented with `TURN ASSIST` on top of `ACRO` but that didn't fly nice at all. It felt like INAV was constantly trying to increase the bank angle when making turns, giving a very unstable feeling. That brings me to the question of how `TURN ASSIST` is changing the roll rate. In the example of `ACRO`, I thought it worked like this: pilot puts the airplane at a certain bank angle -> `TURN ASSIST` applies the appropriate up elevator and rudder to make a coordinated turn. Where in this procedure does the roll rate need to be changed for coordinated turns? Shouldn't the roll rate be zero?

From https://github.com/iNavFlight/inav/blob/ff2b2a5447da7d794f94d390ac20e93c8e693585/src/main/flight/pid.c#L892 I get that the roll target gets updated as well.

**Comment by digitalentity** [comment]
When doing a turn, we want plane to rotate around the gravity vector (vertical relative to Earth). This rotation translates to a rotation around all 3 axis in airplane-relative frame.

Imagine a few edge cases:

1. Plane is flat (zero pitch, zero roll): Turn in Earth frame results is Yaw in airplane frame.
1. Plane is banked 90 degrees (pitch = 0,  roll = 90): Turn in Earth frame results is now Pitch in airplane frame.
1. Plane is diving 90 degrees (pitch = 90,  roll = 0): Turn in Earth frame results is now Roll in airplane frame.

All cases of 0..90 deg roll and pitch will translate to non-zero yaw, roll and pitch rotations.

I'm very much against over-complicating the logic by introducing arbitrary additional gains, unless we have a solid physical+mathematical reason to do so.

I think this PR needs to be deprecated in favor of improving the calculation of desired turn rate [here](https://github.com/iNavFlight/inav/blob/master/src/main/flight/pid.c#L853-L874)

**Comment by avs2** [comment]
Okay, I understand now. I appreciate you taking the time to explain it. 

> I'm very much against over-complicating the logic by introducing arbitrary additional gains, unless we have a solid physical+mathematical reason to do so.

I can now think of a possible explanation for my pitch up problem. When the plane is banked 45 degrees with the nose pointed at the horizon (pitch 0, roll 45), a turn in the earth reference frame is a combination of yaw and pitch in the airplane reference frame. Thus, TURN ASSIST gives some up elevator and rudder in the direction of the turn. But flying wings have no yaw control so that just leaves the up elevator command, causing the nose to point above the horizon and the plane to gain altitude. If it would have had yaw authority, adding yaw input would effectively push the nose down, balancing the nose up effect of the up elevator. Does that make sense?

> I think this PR needs to be deprecated in favor of improving the calculation of desired turn rate here
I don't think the turn rate calculation can be improved. It is just the way the outcome is handled by the pid controller that can cause problems imo.

P.S.: Pushing the commit that implements the roll gain was a mistake on my part (new to git). It was just meant for myself to experiment with completely turning off `TURN ASSIST` in the navigation modes.

**Comment by digitalentity** [comment]
Hm... good catch re: no yaw authority on wings :+1:

In this case pitch gain comes in quite handy. Wings will still need pitching up to prevent loosing altitude in a turn, but not as much as classic airplanes as they don't have a rudder to compensate for excessive pitching up.

**Comment by avs2** [comment]
Just thinking it through, would it make sense so set the roll gain by default to 0.5 for airplanes without yaw control? 0.250 was too high for my Dart 250 and 0 flew best, so obviously it is very airplane dependent.

**Comment by avs2** [comment]
I'm sorry I don't know what I am doing with git. I tried to update the master branch in my fork from the base repo by using a pull request, rebased my branch of the new master and pushed it back to remote (because Sourcetree prompted me to). I hadn't realized this would be visible here.

**Comment by avs2** [comment]
I think this can be merged as I have had several flights with my Dart 250 with `fw_turn_assist_pitch_gain = 0.1` and encountered no issues.

**Comment by Mateyhv** [comment]
@avs2 does the '''fw_turn_assist_pitch_gain = 0.1''' increases or reduces the calculated pitch during a turn? From your initial post I understood that you were experiencing pitch up problems during navigation mode turns which I find strange because normal aerodynamic behaviour is pitching down (test it in manual mode) during a turn.

**Comment by avs2** [comment]
@Mateyhv Not many people seem to know this but in the the navigation modes `TURN ASSIST` is always active on fixed wing. This adds up elevator, and rudder if the airplane has it, to make a coordinated turn and to keep the nose from pitching down as you say. I found this to apply too much up elevator on my Dart 250, causing it to picth up. `fw_turn_assist_pitch_gain` < 1 deceases the calculated pitch, >1 increases it, akin to `fw_turn_assist_yaw_gain` for rudder.

**Comment by Mateyhv** [comment]
@avs2 good clarification, thank you! This was what I was looking for as obviously there was something wrong with the navigation algorythm! TBH I didn't expect the ```TURN ASSIST``` to be always active since flying wings do not benefit from a control surface they don't have which in turn has an impact on their pitch behaviour. 

Isn't a good enough solution to deactivate ```TURN ASSIST``` by default for flying wings instead of adding the new variable? Don't get me wrong you have worked hard to find the solution and implement it and thats great, just trying to simplificate things (whenever possible of course).

**Comment by avs2** [comment]
Well as you said, without stabilization the wing would pitch down when banked in a turn. Of course, the rate, angle and altitude controllers will fight this movement, but they only do so after the error occurs. `TURN ASSIST` tries to avoid the error instead, which I think is a nice feature. 

As you can see in the discussion in this PR and also in #5800 I have had my fair share of issues with `TURN ASSIST`, but I don't see any obvious alternatives. The main issue imo is that it works well in theory, but in reality there are always factors (no yaw control, rudder affecting roll, etc.) causing airplanes to behave differently in the air than INAV expects. However for most setups the current behavior of `TURN ASSIST` works fine, or at least good enough. Until there is a fundamental improvement of the logic that addresses these issues, `fw_turn_assist_pitch_gain` allows for a direct fix of one of them. 

That said, setting `fw_turn_assist_pitch_gain = 0` and `fw_turn_assist_roll_gain = 0` (now removed from this PR), effectively disables `TURN ASSIST` on a wing.

**Comment by Mateyhv** [comment]
I see. So basicaly on a flying wing ```TURN ASSIST``` is overpushing through the pitch channel (because it is actualy doing the math for a coordinated turn on a tailed a/c) so that PI has to correct backing up and you added a gain to tame out the pitch channel push.

On a tailed a/c the coordinated turn is achieved through the rudder pushing the nose into the oncoming trajectory but on a wing the only goal is to stay level into the turn. Since their flying dynamics are quite different shouldn't be better to have different calculation methods for a tailed a/c and a wing on the turn assist function? That doesn't rule out the need of a gain to fine tune for each model of course.

**Comment by digitalentity** [comment]
@Mateyhv fair point. Do you know what the math for the flying wing should be? In the meanwhile I'm fine with an extra gain to reduce the pitch rate for a wing.

**Review by digitalentity — APPROVED** [review]

**Comment by digitalentity** [comment]
I'm merging this. As long as we keep the default to have the same behavior - this should be acceptable until we have a better solution for rudderless aircrafts.

**Comment by b14ckyy** [comment]
The default value needs to be lowered. I tested this with the 2.6 Alpha2 from Pavel (GCC-9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599]) and it caused my AR Wing Pro to stall everytime I did a tight turn in 3D cruise and also in a turn for RTH. O lowered the gain to 0.1 and not the pitch is rock solid pointing at the horizon no matter how fast I turn. 

Behavior with default 1.0 gain: https://youtu.be/zI7pLo4DeDk
Behavior with 0.1 gain: https://youtu.be/YpLyBQrFYjQ

@digitalentity recommended to include this value to the presets table to have it default on 0.1 on Wings and 1.0 on Planes. IMHO it might be more secure to have the overall default very low. A too low value won't harm but a too high value can stall the plane and lead to a crash.

**Comment by digitalentity** [comment]
The problem is that the default value is different on rudder-less wings and classic airplanes. On classic airplanes a gain of 1.0 is required to fly properly, without rudder counteracting the turn. Therefore we can't change the firmware-wide default and should change the value applied via presets (via Configurator)

**Pull request merged**